### DISCUSSION TITLE
Style-spec slots

### DISCUSSION THREAD

**Discussion by davidemerritt** [maplibre/maplibre-gl-js#3787] [Ideas]
I was wondering if any discussion had occurred about supporting slots https://docs.mapbox.com/style-spec/reference/slots/.

This would be a really useful tool to help us with layer ordering as we currently have to render all layers (lets say 1-10) - possibly with 0 data - because they are toggleable by the user, to maintain the correct z-index.

**Comment by HarelM** [comment]
Not that I'm aware of, generally speaking, we can't copy from mapbox so I wouldn't know how to reconcile this...

**Reply by davidemerritt** [reply]
Is there a way to create a feature request, or is this the best forum for that. 

Our use case is explicitly ordered layers that the end user can toggle on/off. 

With the current `beforeId` attribute, each layer wouldneed to understand the other layers (which is tough from a soft architecture perspective). Without using slots or an index approach we have to maintain all of the layers and use {layout: {visibility: 'none'}} to hide the ones that have been toggled off, which seems a bit of an awkward solution, but it does work.

**Reply by wipfli** [reply]
If your main use case is to toggle the visibility of multiple layers, we might also just think about introducing global variables that can be used by multiple layers...

**Reply by boeckMt** [reply]
@davidemerritt we have implemented something that might help you, but it is a bit customized for our use case...

We have some generic layer objects that serve as pointers to the layers of maplibre (based on their IDs). If you change the order of them, you can detect the change and move the layers accordingly.

- [getLayerChangeOrder](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/6255646cf3a0dd9f7e20d27d3872f16aa90922f6/projects/map-maplibre/src/lib/maplibre.helpers.ts#L281)
- [changeOrderOfLayer](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/6255646cf3a0dd9f7e20d27d3872f16aa90922f6/projects/map-maplibre/src/lib/maplibre.helpers.ts#L345C17-L345C35)
- [setVisibility](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/6255646cf3a0dd9f7e20d27d3872f16aa90922f6/projects/map-maplibre/src/lib/maplibre.helpers.ts#L97C17-L97C30)

For `Slots` we have a [similar concept](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/6255646cf3a0dd9f7e20d27d3872f16aa90922f6/projects/map-maplibre/src/lib/maplibre.helpers.ts#L190) like bottom, middle and top [`Filtertypes `](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/6255646cf3a0dd9f7e20d27d3872f16aa90922f6/projects/services-layers/src/lib/types/Layers.ts#L103), but it is based on our generic layer objects, which only serve as containers for different `LayerSpecification`'s, so not the layers inside the `LayerSpecification`

**Reply by pnorman** [reply]
> Not that I'm aware of, generally speaking, we can't copy from mapbox so I wouldn't know how to reconcile this

We can't copy MapBox's implementation - but the idea of putting layers into slots isn't unique to them. We would want a developer to implement it who hasn't looked at the new Mapbox GL source code so they couldn't be tainted by it.

**Reply by mhernz** [reply]
@HarelM \<IANAL>

The Mapbox style spec describes a file format for expressing the content and presentation of an interactive map. 

"Copyright does not protect facts, ideas, systems, or methods of operation" [1], and as such, courts in the US and abroad have held that file formats do not constitute copyrightable expression [2][3].

Therefore it seems to me that MapLibre should be free to provide 1:1 support for Mapbox style properties, so long as the implementations are not derived from Mapbox source code.  

\</IANAL>

[1] https://www.copyright.gov/help/faq/faq-general.html
[2] https://www.lexology.com/library/detail.aspx?g=b5ab25ca-872b-43bf-8040-a5e9e3c94e67
[3] https://law.justia.com/cases/federal/appellate-courts/cafc/21-1542/21-1542-2023-04-06.html

**Reply by mhernz** [reply]
And indeed I think it is strongly in users' interests to support the latest Mapbox spec as closely as possible. How useful would LibreOffice be if it couldn't open .docx files produced by modern versions of MS Word?

**Reply by HarelM** [reply]
A json is a file format, I'm not sure the spec scheme is not copy right protected.
I'm not a lawyer though, and I have little interest in debating about it.
Bottom line, maplibre doesn't aim to chaise mapbox's spec as we simply don't have the resources or motivation to do it.
We have our community and our needs and they drive the changes, this is how we operate.
If you have the resources to allocate for this I'll be happy to review any PR that adds existing mapbox functionality without coping their code.
You can also just create a conversion tool from mapbox json to maplibre json, it's probably a lot simpler than anything else.

As to your LibreOffice referece, **I** think LibreOffice is a good example of something that is not useful enough, probably for the same resources reason, they can't chaise Microsoft with new features etc, so they can offer an editor that might support some docx features. The closest thing is google docs, which have google resources, so they might be able to do something functional, but still probably can convert a lot of stuff that exists in a docx file...

In any case, feel free to submit a PR or open a spec change proposal in the spec repo, it will be much more productive :-)

**Comment by j-osephlong** [comment]
I understand that we can't backport, but would it not be possible to develop something that more or less achieves the same thing without infringing on mapbox? One idea could be add a parameter to the add layer object, call it 'region', where one supplies a name for a layer stack region. The ordering of regions themselves could be first come first serve, or there could be a sort of regions definition dict in the map initialization parameters. This seems to solve the same thing, more or less. I know the implications of this would be widespread throughout the code base, but I'm just wondering if this approach would be on the same page with the maintainers of maplibre.

**Reply by wipfli** [reply]
You can achieve the functionality of slots with hidden background layers, no?

```json
{
  "id": "my-slot",
  "type": "background",
  "layout": {"visibility": "none"}
}
```

**Reply by j-osephlong** [reply]
Is the idea here to use the empty layer in the beforeId parameter? That's pretty clever.

**Comment by pnorman** [comment]
I was searching for maplibre slots and it gave this somewhat amusing AI overview

> In MapLibre GL, "slots" are predefined placeholder locations within a map style's layer stack that allow users to insert custom layers at specific, stable points without needing to reference individual basemap layer IDs. Defined by style authors using a "slot" layer in the style specification, they provide well-defined insertion points (like "above-polygons" or "above-poi") that ensure custom layers are positioned correctly relative to the basemap's elements, even as the basemap updates
